### PR TITLE
add python script to get rid of empty ntuples in dataset xmls

### PR DESCRIPTION
**Pull request by MatthiesC**
see *documentation* in code for further details

### CODE REVIEW

**Comment by raggleton** [comment]
Thanks for this, do you have any examples of empty ntuples on /nfs to test this on? In _theory_ sframe first scans over all the input files in the XML and figures out which one(s) to skip, but clearly something is going awry

To tackle the actual underlying issue, do you have the crab logs of jobs that produce empty files? (or which dataset...can try to run the jobs again myself).  Would be good to nail down why this is happening

**Comment by MatthiesC** [comment]
To test my script, you can run it with this exemplary dataset xml:
`/nfs/dust/cms/user/matthies/Analysis_80x_v5/CMSSW_8_0_24_patch1/src/UHH2/scripts/crab/Robin/SingleElectron_D_rest.xml`
(Copy it first into one of your directories)
To view the sizes of the ntuples linked in that file, look at the output of (last five files listed):
`ls -lh -S /pnfs/desy.de/cms/tier2/store/user/cmatthie/RunII_80X_v5/SingleElectron/crab_SingleElectron_D_rest/180806_123824/0000/`
As you can see, there are five ntuples which are empty (size = 36K). After running my script, these five ntuples are commented out within the new dataset xml. You can run readaMCatNloEntries.py on both xmls to verify that both give the same number of events since just empty ntuples were removed.

As far as I can tell, empty ntuples are just produced when processing real data and not MC. But I could be wrong. A possible reason for this is that events are distributed significantly more inhomogenously for real data than for MC over the CRAB jobs.

You can find the logs for the above exemplary CRAB job at `http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=Christopher+Matthies&refresh=0&table=Jobs&p=1&records=100&activemenu=2&status=finished&site=&tid=180806_123824%3Acmatthie_crab_SingleElectron_D_rest`
The job/postjob logs of job 19, 30, 31, 36, 44 are the interesting ones.

**Comment by raggleton** [comment]
Ah great, thanks for pinpointing jobs where this happens. I just tried the input file from job 19 locally, and it worked OK. To check what crab was actually running, could you please do `crab preparelocal --jobid=19` on that CRAB project dir? This should reproduce the lumi JSON, which I have a sneaky suspicion has no lumi sections for this job.

File: scripts/crab/commentOutEmptyNtuples.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+# Creates: ../../common/datasets/DATA_SingleElectron_RunB_withoutEmptyNtuples.xml
+
+
+import sys
+import os
+
+def read_xml(xmlFileDir):
+    xmlFile = open(str(xmlFileDir))
+    rootFileStore = []
+    comment = False
```
Comment (raggleton) [inline scripts/crab/commentOutEmptyNtuples.py]: This bool doesn't seem to get used? Only set in the `if ` statements.

**Comment by MatthiesC** [comment]
The crab.log file after running job 19 locally (I hope this is what you asked for):
`/nfs/dust/cms/user/matthies/Analysis_80x_v5/CMSSW_8_0_24_patch1/src/UHH2/scripts/crab/crab_SingleElectron/crab_SingleElectron_D_rest/crab.log`

**Comment by raggleton** [comment]
Ah I meant all the files crab uses for that job - if you look inside the log file, you'll see it uses "job_lumis_19.json" as a lumi mask. That's the file I'm most interested in. The `crab preparelocal` command should be able to recreate this file.

File: scripts/crab/commentOutEmptyNtuples.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    for line in xmlFile:
+        if '<!--' in line and not '-->':
```
[the remark this answers is not in the document]
Comment (MatthiesC) [reply scripts/crab/commentOutEmptyNtuples.py]: I just blindly copied this read_xml function from readaMCatNloEntries. Committed an improved function for readaMCatNloEntries and commentOutEmptyNtuples

**Comment by raggleton** [comment]
It's ok, I managed to recreate your crab project area (with `crab remake`). Then I used `crab report -d crab_SingleElectron_D_rest`, which gives you lists of lumis it processed, as well as the lumi JSONs for each job (inside `results/run_and_lumis.tar.gz`). From this, I looked at `job_lumis_19.json`. This has contents:

```
{"276495": [[82, 85]]}
```

If you then rerun the ntuplewriter with that lumi mask, it will do exactly as in the crab job, and not run over any events. 

So this isn't an issue with our code it appears, rather that the crab splitting was a bit odd. Which splitting method did you use in your crab config? Also, did you apply a JSON mask in the crab config?

**Comment by MatthiesC** [comment]
I used `config.Data.splitting = 'EventAwareLumiBased'` and applied a lumi mask via JSON file (the leftovers of cmatthie_crab_SingleElectron_D), yes.

File: scripts/crab/commentOutEmptyNtuples.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+def commentOutEmptyRootFiles(xmlfiles):
+    print 'Begin commenting out empty Ntuples ...'
+    emptyRootFiles_allxmls = []
+    for xml in xmlfiles:
+        rootFileStore = read_xml(xml)
+        emptyRootFiles = findEmptyRootFilesBySize(rootFileStore)
+        emptyRootFiles_allxmls.append(emptyRootFiles)
+        xmlFile_in = open(str(xml))
+        xmlFile_out_path = str(xml)
+        xmlFile_out_path = xmlFile_out_path[0:len(xmlFile_out_path)-4]+'_withoutEmptyNtuples.xml'
```
[the remark this answers is not in the document]
Comment (MatthiesC) [reply scripts/crab/commentOutEmptyNtuples.py]: included in the new commits

File: scripts/crab/commentOutEmptyNtuples.py (hunk 1 above)
Comment (MatthiesC) [reply scripts/crab/commentOutEmptyNtuples.py]: See other message. Now, this boolean has a meaning.

**Comment by raggleton** [comment]
OK - according to the crab docs, `LumiBased` or `automatic` are recommended for data (whether that would have avoided this, I don't know). It seems odd that `EventAwareLumiBased` doesn't "know" that those particular lumisections have no events, given it aims to have a roughly constant # events / job...

**Comment by raggleton** [comment]
More interesting: I wondered if there were any events in those LS. Using brilcalc:

```
brilcalc lumi -i "{276495: [[82, 85]]}" --byls
#Data tag : v2 , Norm tag: onlineresult
+-------------+-------+-------------------+--------------+--------+----------------+---------------+-------+--------+
| run:fill    | ls    | time              | beamstatus   | E(GeV) | delivered(/ub) | recorded(/ub) | avgpu | source |
+-------------+-------+-------------------+--------------+--------+----------------+---------------+-------+--------+
| 276495:5078 | 82:82 | 07/07/16 11:19:01 | ADJUST       | 6500   | 225950.334     | 219232.778    | 31.6  | HFOC   |
| 276495:5078 | 83:83 | 07/07/16 11:19:24 | ADJUST       | 6500   | 242004.495     | 234397.828    | 35.8  | HFOC   |
| 276495:5078 | 84:84 | 07/07/16 11:19:48 | ADJUST       | 6500   | 241918.263     | 234364.569    | 35.8  | HFOC   |
| 276495:5078 | 85:85 | 07/07/16 11:20:11 | STABLE BEAMS | 6500   | 242088.974     | 234561.140    | 35.8  | HFOC   |
+-------------+-------+-------------------+--------------+--------+----------------+---------------+-------+--------+
```

So it seems that only LS 85 is stable beams and has luminosity.

But if you look inside the Golden JSON, for that run in question, only `"276495": [[87, 268]],` appears. So those LS 82-85 are not Golden (presumably because they don't contain anything?). It's just unfortunate that they got put in their own job (presumably CRAB looks at the lumi and divides it up, irrespective of beamstatus?) 

I don't exactly know why this has never been an issue before... maybe some magic combination of your Lumi mask, splitting mode, and your unitsPerJob setting?

**Comment by raggleton** [comment]
BTW script looks good, thanks 👍

**Comment by dgonzal** [comment]
hey guys,

checking the size of the files is potentially very dangerous, since the size depends on our ntuple setup and can be very large or very small.  

Cheers

**Comment by raggleton** [comment]
Hm yes good point. I _think_ that doing `AnalysisTree.GetEntries() == 0` should be a better check? My worry was that the job might not write out a Tree at all (which is actually another check that should be done in terms of "bad" ntuples)

**Comment by raggleton** [comment]
So checks that should be done on each ROOT file in the script:

- Check input ROOT file actually exists: `os.path.isfile(...)`

- Check file is valid ROOT file: use `f.IsZombie()`

- If the TTree actually exists in the file: `tree = f.Get("AnalysisTree")`, `if tree != None` ... (could also cast the object to a bool & test)

- If the TTree is empty: `tree.GetEntries() == 0`

I think this should cover all the various ways a file can go bad.
Might also be worth adding in a printout of "bad" files such that the user is aware that some of their ntuples are invalid, rather than silently hiding them, since bad ntuples are not expected.

@MatthiesC would you be able to implement these checks?

**Comment by MatthiesC** [comment]
I should be able to do this. But I will keep the "fast" method of just checking the size as an option I guess.

**Comment by dgonzal** [comment]
This should be safe. Nonetheless it might be rather slow since you might have to open a few thousand files. To check them in parallel would be nice and there is already some examples e.g. 

UHH2/scripts/crab/readaMCatNloEntries.py 

If you do it, you should take care to join the processes before commenting the lines in one go. If you need a hand let me know.

**Comment by MatthiesC** [comment]
I think the most efficient way is a hybrid of both: Check the file sizes first and then open just the ones which fall into the category "very small" (e. g. < 1 MB or something). This should be way less time and ressource consuming. Of course, this can then still be done in parallel on multiple cores. I will see.

**Pull request merged**